router/core/http_server.go

```diff
@@ -0,0 +1,134 @@
+package core
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"github.com/wundergraph/cosmo/router/pkg/health"
+	"go.uber.org/zap"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type server struct {
+	mu          sync.RWMutex
+	httpServer  *http.Server
+	tlsConfig   *TlsConfig
+	logger      *zap.Logger
+	handler     http.Handler
+	healthcheck health.Checker
+	baseURL     string
+	graphServer *graphServer
+}
+
+type httpServerOptions struct {
+	addr            string
+	logger          *zap.Logger
+	tlsConfig       *TlsConfig
+	tlsServerConfig *tls.Config
+	healthcheck     health.Checker
+	baseURL         string
+}
+
+func newServer(opts *httpServerOptions) *server {
+	httpServer := &http.Server{
+		Addr: opts.addr,
+		// Covers the time from when the connection is accepted to when the request body is
+		// fully read (if you do read the body, otherwise to the end of the headers).
+		// It's implemented in net/http by calling SetReadDeadline immediately after Accept.
+		ReadTimeout: 1 * time.Minute,
+		// WriteTimeout normally covers the time from the end of the request header read to the end
+		// of the response write (a.k.a. the lifetime of the ServeHTTP), by calling
+		// SetWriteDeadline at the end of readRequest.
+		WriteTimeout: 2 * time.Minute,
+		ErrorLog:     zap.NewStdLog(opts.logger),
+		TLSConfig:    opts.tlsServerConfig,
+	}
+
+	n := &server{
+		httpServer:  httpServer,
+		tlsConfig:   opts.tlsConfig,
+		logger:      opts.logger,
+		mu:          sync.RWMutex{},
+		healthcheck: opts.healthcheck,
+		baseURL:     opts.baseURL,
+	}
+
+	httpServer.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// Multiple requests can read the handler at the same time, but only one goroutine can write it.
+		// When swapping the graph server there might be in-flight requests that are still being processed
+		// but this is tolerable because we are waiting for them to finish before shutting down the old server.
+		n.mu.RLock()
```

Comment:
Have you thought about this?

```go
n.mu.RLock()
handler := n.handler
n.mu.RUnlock()
handler.ServeHTTP(w, r)
```

This should be memory safe but allows to swap the handler even if traffic is being served on the old handler.
With your approach, it will not be possible to acquire a Lock() while the RLock() was not released, which can take a while.

Reply:
Your assumption doesn't seem correct. The std library says:

> If any goroutine calls Lock while the lock is already held by one or more readers, concurrent calls to RLock will block until the writer has acquired (and released) the lock, to ensure that the lock eventually becomes available to the writer. Note that this prohibits recursive read-locking

When calling Lock() we won't let any new requests obtain a read lock until the write lock was released. Existing RLock() aren't affected. Because we just change a variable, the impact is minimal.

Reply:
Example https://go.dev/play/p/LDyzDIfCoNx

Reply:
Fake your own tests 😄 I was wrong. Good catch!